ipn/auditlog/auditlog.go

```diff
@@ -0,0 +1,473 @@
+	ctxCancel   context.CancelFunc // cancels ctx.
+	backoffOpts                    // backoff settings for retry operations.
+
+	// mu protects the fields below.
+	mu        sync.Mutex
+	store     LogStore      // persistent storage for unsent logs.
+	profileID ipn.ProfileID // empty if [Logger.SetProfileID] has not been called.
+	transport Transport     // nil until [Logger.Start] is called.
+}
+
+// NewLogger creates a new [Logger] with the given options.
+func NewLogger(opts Opts) *Logger {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	l := &Logger{
+		retryLimit:  opts.RetryLimit,
+		logf:        logger.WithPrefix(opts.Logf, "auditlog: "),
+		store:       opts.Store,
+		flusher:     make(chan struct{}, 1),
+		done:        make(chan struct{}),
+		ctx:         ctx,
+		ctxCancel:   cancel,
+		backoffOpts: defaultBackoffOpts,
+	}
+	l.logf("created")
+	return l
+}
+
+// FlushAndStop synchronously flushes all pending logs and stops the audit logger.
+// This will block until a final flush operation completes or context is done.
+// If the logger is already stopped, this will return immediately.  All unsent
+// logs will be persisted to the store.
+func (l *Logger) FlushAndStop(ctx context.Context) {
+	l.stop()
+	l.flush(ctx)
+}
+
+// SetProfileID sets the profileID for the logger. This must be called before any logs can be enqueued.
+// The profileID of a logger cannot be changed once set.
+func (l *Logger) SetProfileID(profileID ipn.ProfileID) error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	if l.profileID != "" {
+		return errors.New("profileID already set")
+	}
+
+	l.profileID = profileID
+	return nil
+}
+
+// Start starts the audit logger with the given transport.
+// Returns an error if the logger is already started.
+func (l *Logger) Start(t Transport) error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if l.transport != nil {
+		return errors.New("already started")
+	}
+
+	l.transport = t
+	pending, err := l.storedCountLocked()
+	if err != nil {
+		l.logf("[unexpected] failed to restore logs: %v", err)
+	}
+	go l.flushWorker()
+	if pending > 0 {
+		l.flushAsync()
+	}
+	return nil
+}
+
+// ErrAuditLogStorageFailure is returned when the logger fails to persist logs to the store.
+var ErrAuditLogStorageFailure = errors.New("audit log storage failure")
+
+// Enqueue queues an audit log to be sent to the control plane (or another suitable consumer/transport).
+// This will return an error if the underlying store fails to save the log or we fail to generate a unique
+// eventID for the log.
+func (l *Logger) Enqueue(action tailcfg.ClientAuditAction, details string) error {
+	txn := &transaction{
+		Action:    action,
+		Details:   details,
+		TimeStamp: time.Now(),
+	}
+
+	// generate a unique eventID for the log. This is used to de-duplicate logs
+	// persisted to the store.
+	bytes := make([]byte, 16)
+	if _, err := rand.Read(bytes); err != nil {
+		return fmt.Errorf("%w: %w", ErrAuditLogStorageFailure, err)
+	}
```

Comment:
This call never returns nil. It panics instead. So you can drop the error handling.